src/Widgets/PowerModeList.vala

```diff
@@ -0,0 +1,174 @@
+    public static bool successfully_initialized { get; private set; default = true; }
+
+    private static Services.DBusInterfaces.PowerProfile? pprofile;
+    private static GLib.Settings? settings;
+
+    public string settings_key { get; construct; }
+
+    private Gtk.RadioButton saver_radio;
+    private Gtk.RadioButton balanced_radio;
+    private Gtk.RadioButton performance_radio;
+
+    private const string ICON_RES = "/io/elementary/desktop/wingpanel/power/scalable/categories/";
+
+    public PowerModeList (bool on_battery) {
+        Object (
+            settings_key: on_battery ? "profile-on-good-battery" : "profile-plugged-in",
+            orientation: Gtk.Orientation.VERTICAL,
+            margin: 6
+        );
+    }
+
+    static construct {
+        var schema = SettingsSchemaSource.get_default ().lookup ("io.elementary.settings-daemon.power", true);
+        if (schema != null && schema.has_key ("profile-plugged-in") && schema.has_key ("profile-on-good-battery")) {
+            settings = new GLib.Settings ("io.elementary.settings-daemon.power");
+        } else {
+            warning ("settings-daemon schema not found, will connect to power-profiles-daemon directly");
+        }
+
+        try {
+            pprofile = Bus.get_proxy_sync (
+                BusType.SYSTEM,
+                Services.DBusInterfaces.POWER_PROFILES_DAEMON_NAME,
+                Services.DBusInterfaces.POWER_PROFILES_DAEMON_PATH,
+                DBusProxyFlags.NONE
+            );
+        } catch (Error e) {
+            critical (e.message);
+            successfully_initialized = false;
+        }
+    }
+
+    construct {
+        if (pprofile == null) {
+            return;
+        }
+
+        build_ui ();
+        build_events ();
+    }
+
+    private void build_ui () {
+        Gtk.IconTheme.get_default ().add_resource_path ("/io/elementary/desktop/wingpanel/power");
+
+        var saver_icon = new Gtk.Image.from_icon_name ("power-mode-powersaver-symbolic", Gtk.IconSize.BUTTON);
+
+        var saver_label = new Gtk.Label (_("Power Saver"));
+
+        var saver_button_box = new Gtk.Box (HORIZONTAL, 6) {
+            halign = Gtk.Align.START,
+            margin = 3
+        };
+        saver_button_box.add (saver_icon);
+        saver_button_box.add (saver_label);
+
+        saver_radio = new Gtk.RadioButton (null);
+        saver_radio.get_style_context ().add_class ("image-button");
+        saver_radio.add (saver_button_box);
+
+        var balanced_icon = new Gtk.Image.from_icon_name ("power-mode-balanced-symbolic", Gtk.IconSize.BUTTON);
+
+        var balanced_label = new Gtk.Label (_("Balanced"));
+
+        var balanced_button_box = new Gtk.Box (HORIZONTAL, 6) {
+            halign = Gtk.Align.START,
+            margin = 3
+        };
+        balanced_button_box.add (balanced_icon);
+        balanced_button_box.add (balanced_label);
+
+        balanced_radio = new Gtk.RadioButton.from_widget (saver_radio);
```

Comment:
Can we use `group` property instead of `from_widget` constructor? There's no `from_widget` in GTK4

Reply:
Done!